File: app/src/routes/xcode.rs
use crate::xcode;
use axum::{http::StatusCode, response::IntoResponse, Json};
use serde::Deserialize;
use serde_json::json;
use std::path::Path;

#[derive(Debug, Deserialize)]
pub struct DiscoverProjectRequest {
    pub path: String,
}

#[derive(Debug, Deserialize)]
pub struct BuildSchemeRequest {
    pub path: String,
    pub scheme: String,
}

#[derive(Debug, Deserialize)]
pub struct GetLaunchableProductsRequest {
    pub build_dir: String,
}

/// Discover Xcode project information (schemes, targets, configurations)
pub async fn discover_project(Json(request): Json<DiscoverProjectRequest>) -> impl IntoResponse {
    let path = Path::new(&request.path);

    match xcode::discover_project(path).await {
        Ok(project) => {
            (StatusCode::OK, Json(serde_json::to_value(project).unwrap())).into_response()
        }
        Err(error) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "error": error.to_string() })),
        )
            .into_response(),
    }
}

/// Build an Xcode scheme for iOS Simulator with code signing disabled
pub async fn build_scheme(Json(request): Json<BuildSchemeRequest>) -> impl IntoResponse {
    let path = Path::new(&request.path);

    match xcode::build_scheme(path, &request.scheme).await {
        Ok(result) => (StatusCode::OK, Json(serde_json::to_value(result).unwrap())).into_response(),
        Err(error) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "error": error.to_string() })),
        )
            .into_response(),
    }
}

/// Get launchable products from a build directory
pub async fn get_launchable_products(
    Json(request): Json<GetLaunchableProductsRequest>,
) -> impl IntoResponse {
    match xcode::get_launchable_products_from_dir(&request.build_dir).await {
        Ok(products) => {
            (StatusCode::OK, Json(serde_json::to_value(products).unwrap())).into_response()
        }
        Err(error) => (
            StatusCode::BAD_REQUEST,
            Json(json!({ "error": error.to_string() })),
        )
            .into_response(),
    }
}

File: app/src/main/routes/xcode.ts
import { Hono } from 'hono';
import { streamSSE } from 'hono/streaming';
import { discoverProject, buildSchemeStream, getLaunchableProducts } from '../services/xcode';

export const xcodeRoutes = new Hono();

xcodeRoutes.post('/discover', async (c) => {
  const body = await c.req.json();
  const projectPath = body.path;

  if (!projectPath) {
    return c.json({ error: 'Path is required' }, 400);
  }

  try {
    const project = await discoverProject(projectPath);
    return c.json(project);
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return c.json({ error: message }, 400);
  }
});

xcodeRoutes.post('/build/stream', async (c) => {
  const body = await c.req.json();
  const projectPath = body.path;
  const scheme = body.scheme;

  if (!projectPath || !scheme) {
    return c.json({ error: 'Path and scheme are required' }, 400);
  }

  return streamSSE(c, async (stream) => {
    const emitter = buildSchemeStream(projectPath, scheme);

    const onEvent = async (event: unknown) => {
      await stream.writeSSE({ data: JSON.stringify(event) });
    };

    emitter.on('event', onEvent);

    await new Promise<void>((resolve) => {
      emitter.on('end', () => {
        resolve();
      });

      // Handle client disconnect
      stream.onAbort(() => {
        emitter.removeAllListeners();
        resolve();
      });
    });
  });
});

xcodeRoutes.get('/launchable-products', async (c) => {
  const buildDir = c.req.query('buildDir');

  if (!buildDir) {
    return c.json({ error: 'buildDir is required' }, 400);
  }

  try {
    const products = await getLaunchableProducts(buildDir);
    return c.json({ products });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return c.json({ error: message }, 500);
  }
});
